### ISSUE TITLE
token 3.1.0 behaviour difference to token 2.6.0

### ISSUE THREAD

**Issue by mvines** [solana-labs/solana-program-library#1231]
The mainnet-beta transaction [cgq..Fgc](https://explorer.solana.com/tx/cqgPsDCZNLd45A8LZC7PE7i646XYqgx6xdXSLGyHS9ba2QXkfEc1FrchBBVm6Dgt6xcjgPxLsqfgYjd19KMQFgc) succeeded, but failed on the `bvb` running a custom build that shims in the token 3.1.0 binary.

Transaction details from `bvb` that caused the node to fork off mainnet-beta:
```
      Recent Blockhash: 8ZYNc26JJwgTnNjQ9kfZSUpugqjBBFhRiSQVEYqdeAM1
      Signature 0: cqgPsDCZNLd45A8LZC7PE7i646XYqgx6xdXSLGyHS9ba2QXkfEc1FrchBBVm6Dgt6xcjgPxLsqfgYjd19KMQFgc
      MessageHeader { num_required_signatures: 1, num_readonly_signed_accounts: 0, num_readonly_unsigned_accounts: 2 }
      Account 0: 5Ktk6Ci36W7hVQYsDWggHRhgsTdmtSGxCXf35Y5fsGDK
      Account 1: 8em6NLLarGyvsAvYTq93Kyip8vvWt7mm9S6kkwKBpGQx
      Account 2: kinXdEcpDQeHPEuQnqmUgtYykqKGVFq6CeVX5iAHJq6
      Account 3: TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA
      Instruction 0
        Program: TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA (3)
        Account 0: 8em6NLLarGyvsAvYTq93Kyip8vvWt7mm9S6kkwKBpGQx (1)
        Account 1: kinXdEcpDQeHPEuQnqmUgtYykqKGVFq6CeVX5iAHJq6 (2)
        Account 2: 8em6NLLarGyvsAvYTq93Kyip8vvWt7mm9S6kkwKBpGQx (1)
        Account 3: 5Ktk6Ci36W7hVQYsDWggHRhgsTdmtSGxCXf35Y5fsGDK (0)
        Data: [12, 224, 216, 106, 58, 154, 2, 0, 0, 18]
      Status: Error processing Instruction 0: custom program error: 0x12
        Fee: ◎0.000005
        Account 0 balance: ◎1 -> ◎0.999995
        Account 1 balance: ◎0.00203928
        Account 2 balance: ◎0.0014616
        Account 3 balance: ◎1.1305824
      Log Messages:
        Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [1]
        Program log: Instruction: TransferChecked
        Program log: Error: decimals different from the Mint decimals
        Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 3492 of 200000 compute units
        Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA failed: custom program error: 0x12
```

**Comment by mvines** [comment]
I think this is probably ok.   Previously the self-transfer was a blind ✔️, but now with 3.1.0 we actually fully validate the transfer, so this transaction legitimately should have failed.    If you agree Trent, let's chat with the kin team about it

**Comment by t-nelson** [comment]
> I think this is probably ok. Previously the self-transfer was a blind heavy_check_mark, but now with 3.1.0 we actually fully validate the transfer, so this transaction legitimately should have failed. If you agree Trent, let's chat with the kin team about it

Yep this failure is expected, 18 != 5.  I'll raise with KIN folks tomorrow

**Comment by mvines** [comment]
Found another invalid transaction: [2x4...TwU](https://explorer.solana.com/tx/2x4cPZauyZuCih3b8zYfF61B84P1U3HHD5gdFbL9zyeCvnohSyqvkzkyMi7p8qASFvAVyBqUotkt9PKnF1jhNTwU)

This was a self-transfer of 550 KIN when the account only had 540 KIN.

3.1.0 correctly rejects this transaction with `TokenError::InsufficientFunds`: 
> Status: Error processing Instruction 1: custom program error: 0x1

**Comment by mvines** [comment]
A non-kin one this time, a self TransferChecked of 1 wrapped SOL with an account holding 0 SOL: [4Qu...goD](https://explorer.solana.com/tx/4QuWEuqWvDQMniG8bKJxQQaZ3k2Gp6HECUeGHKifGesGGrE5WGaVtNQ5A2iDJA8aZmcG67A1E2bTMawS1Eto8goD).   

3.1.0 correctly rejects this transaction with TokenError::InsufficientFunds:
> Status: Error processing Instruction 1: custom program error: 0x1

**Comment by t-nelson** [comment]
Thankfully looks like none of these have been CPI'd so far, so should all be easily fixable client-side changes :crossed_fingers:

**Comment by mvines** [comment]
100KIN self-transfer between an account with 0 KIN in it: https://explorer.solana.com/tx/27AuGC8YVJbrNAKrtTAG1o6iBNsyrQQs6XskM1nv5PwX7X5fnFtYhKwhMzJJ3kDXazdxDuzyfoCu9UySyKwf3zzm
